app.py

```diff
@@ -14,10 +16,15 @@
 
 class DbpediaWrapper(ClamsApp):
 
-    def __init__(self):
+    def __init__(self, address="localhost:2222"):
         super().__init__()
-        self.address = "http://localhost:2222/rest/annotate"
+        self.address = "http://{}/rest/annotate".format(address)
         self.reqheaders = {'Accept': 'application/json'}
+        self.session = requests.Session()
+        self.session.headers.update(self.reqheaders)
+        # ensure that the server is online
+        r = self.session.get(url=self.address)
```

Comment:
So in the docker setting, this line is called almost immediately after dbps server start-up command is executed. Hence this `GET` requests is a sure-miss as the dbps server would be still starting up. And in fact, I received the following python exception; 
```
requests.exceptions.ConnectionError: HTTPConnectionPool(host='localhost', port=2222): Max retries exceeded with url: /rest/annotate (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f914a1aef70>: Failed to establish a new connection: [Errno 111] Connection refused'))
java -Dfile.encoding=UTF-8 -Xmx10G -jar /dbps/rest/target/rest-1.1-jar-with-dependencies.jar /dbps/en http://0.0.0.0:2222/rest
```

And then the `app.py` just terminates because of this exception. So the code should be something like 

```python 
stat_code = 500
timer = 0
while 200<= stat_code <= 400 and timer <= timeout:
    try:
        r = self.session.get(...)
        stat_code = r.status_code
        timer += interval
        time.sleep(interval)
    except requests.exceptions.ConnectionError:
        timer += interval
        time.sleep(interval)
```

Or using `requests.adapters` (https://requests.readthedocs.io/en/latest/user/advanced/#example-automatic-retries)

``` python
# experiment with `options_to_wait` and find the working combination
retry_adapter = HTTPAdapter(max_retries=Retry(options_to_wait))
self.session.mount('http://', retry_adapter)
self.sessions.get(...)
```
    
Another issue with my original proposal was to put this "wait" feature in the `__init__` method. Thus, while the dbps server is loading up, the `DbpediaWrapper` instance does not fully initiated, and in turn, any API call to the `DbpediaWrapper` instance simply fails. So in fact, what should be done, I think, is 
1. setting up "retrier" and mount to the `self.session` occurs once in the `__init__`. This shouldn't take more that a split second, so the app instance must be fully initiated and its APIs are ready to receive HTTP requests. 
2. actually "waiting" with the retry options should occur inside `_annotate` method. So that when the very first `POST` request is received before the dbps server is fully loaded up, it should wait some time then.

Reply:
> And then the app.py just terminates because of this exception. So the code should be something like

> ```
> stat_code = 500
> timer = 0
> while 200<= stat_code <= 400 and timer <= timeout:
>     try:
>         r = self.session.get(...)
>         stat_code = r.status_code
>         timer += interval
>         time.sleep(interval)
>     except requests.exceptions.ConnectionError:
>         timer += interval
>         time.sleep(interval)
> 
> ```

I think I'm confused. How does this while loop execute if the `stat_code` is set to 500? Wouldn't it need to be between 200 and 400?

> Or using requests.adapters (https://requests.readthedocs.io/en/latest/user/advanced/#example-automatic-retries)

> ```
> # experiment with `options_to_wait` and find the working combination
> retry_adapter = HTTPAdapter(max_retries=Retry(options_to_wait))
> self.session.mount('http://', retry_adapter)
> self.sessions.get(...)
> ```

You said `__init__` should not call `.get()`. So should the first two lines be called in the `__init__` and the third be used in the `_annotate()` method (along with `wait_for_resource()`)?

Reply:
@keighrim I pushed some code, but I don't think it's fixed yet as I'm still getting the same error. Would you mind giving some feedback please?